namer/curator/src/main/scala/io/buoyant/namer/curator/CuratorNamer.scala

```diff
@@ -0,0 +1,124 @@
+package io.buoyant.namer.curator
+
+import com.twitter.finagle._
+import com.twitter.logging.Logger
+import com.twitter.util._
+import org.apache.curator.framework.state.{ConnectionState, ConnectionStateListener}
+import org.apache.curator.framework.{CuratorFramework, CuratorFrameworkFactory}
+import org.apache.curator.retry.ExponentialBackoffRetry
+import org.apache.curator.x.discovery.details.ServiceCacheListener
+import org.apache.curator.x.discovery.{ServiceDiscovery, ServiceDiscoveryBuilder, ServiceInstance}
+import scala.collection.JavaConverters._
+
+/**
+ * The CuratorNamer takes paths of the form
+ *
+ * {{{
+ * /<servicename>/<residual>*
+ * }}}
+ *
+ * And uses the Curator discovery library to resolve that service name.
+ */
+class CuratorNamer(zkConnect: String, basePath: String, idPrefix: Path) extends Namer {
+
+  private[this] val log = Logger.get(getClass.getName)
+
+  private[this] val curator: Future[CuratorFramework] = {
+    val client = CuratorFrameworkFactory.newClient(zkConnect, new ExponentialBackoffRetry(1000, 3))
+    client.start()
+
+    val promise = new Promise[CuratorFramework]()
+    val listenable = client.getConnectionStateListenable
+    listenable.addListener(new ConnectionStateListener {
+      override def stateChanged(
+        client: CuratorFramework,
+        newState: ConnectionState
+      ): Unit = if (newState.isConnected) {
+        log.info("Curator connection established")
+        promise.setValue(client)
+        listenable.removeListener(this)
+      }
+    })
+
+    promise
+  }
+
+  private[this] val serviceDiscovery: Future[ServiceDiscovery[Void]] =
+    curator.map { client =>
+      val disco = ServiceDiscoveryBuilder.builder(classOf[Void]).basePath(basePath)
+        .client(client).build
+      disco.start()
+      log.info("Curator service discovery started")
+      disco
+    }
```

Comment:
Can we make one further small change here:

I think both `curator` and `serviceDiscovery` should be declared as `private[this] lazy val` so that network ops are not initialized during instantiation.

As it is, running `linkerd-examples/test` in SBT without a local zookeeper running causes a bunch of warnings:
```
WARN 1101 20:15:28.993 pool-92333-thread-5-ScalaTest-running-ExamplesTest-SendThread(localhost:2181): Session 0x0 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
	at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:356)
	at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1192)
WARN 1101 20:15:30.100 pool-92333-thread-5-ScalaTest-running-ExamplesTest-SendThread(localhost:2181): Session 0x0 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
	at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:356)
	at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1192)
WARN 1101 20:15:31.203 pool-92333-thread-5-ScalaTest-running-ExamplesTest-SendThread(localhost:2181): Session 0x0 for server null, unexpected error, closing socket connection and attempting reconnect
java.net.ConnectException: Connection refused
	at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method)
	at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716)
	at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:356)
	at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1192)
```

(and there are logged indefinitely).

Other namers avoid network initialization until it is needed.  Making these vals _lazy_ should fix this.